Source/WebCore/platform/graphics/cg/PathCG.cpp

```diff
@@ -445,6 +520,52 @@ FloatRect PathCG::strokeBoundingRect(const Function<void(GraphicsContext&)>& str
     return zeroRectIfNull(box);
 }
 
+static inline void addToCGContextPath(CGContextRef context, PathDataLine segment)
+{
+    addToCGContextPath(context, PathMoveTo { segment.start });
+    addToCGContextPath(context, PathLineTo { segment.end });
+}
+
+static inline void addToCGContextPath(CGContextRef context, PathDataQuadCurve segment)
+{
+    addToCGContextPath(context, PathMoveTo { segment.start });
+    addToCGContextPath(context, PathQuadCurveTo { segment.controlPoint, segment.endPoint });
+}
+
+static inline void addToCGContextPath(CGContextRef context, PathDataBezierCurve segment)
+{
+    addToCGContextPath(context, PathMoveTo { segment.start });
+    addToCGContextPath(context, PathBezierCurveTo { segment.controlPoint1, segment.controlPoint2, segment.endPoint });
+}
+
+static inline void addToCGContextPath(CGContextRef context, PathDataArc segment)
+{
+    addToCGContextPath(context, PathMoveTo { segment.start });
+    addToCGContextPath(context, PathArcTo { segment.controlPoint1, segment.controlPoint2, segment.radius });
+}
+
+static inline void addToCGContextPath(CGContextRef context, PathSegment anySegment)
+{
+    WTF::switchOn(WTFMove(anySegment).data(),
+        [&](auto&& segment) {
+            addToCGContextPath(context, WTFMove(segment));
+        });
+}
+
+void addToCGContextPath(CGContextRef context, const Path& path)
+{
+    if (auto* singleSegment = path.singleSegmentIfExists(); LIKELY(singleSegment)) {
+        addToCGContextPath(context, *singleSegment);
+        return;
+    }
+    if (auto* segments = path.segmentsIfExists(); LIKELY(segments)) {
```

Comment:
What is the purpose of this pattern `LIKELY(segments)`? And how can it be likely after checking `LIKELY(singleSegment)` above. How can this help branch prediction?

Reply:
This function is called for the purpose of it to use the segments, put the segments directly to the CGContext. This function would not be called if we didn't expect most cases to have the segments.

First likely:
It is more likely that the path has single segment than that it does not have.

Second likely, in case the path did not have a single segment:
It is more likely that the path has segments than that it does not have.

```
if (LIKELY(a))
    return "a";
if (LIKELY(b))
   return "b";
```
In this pattern, the "LIKELY" parts do not relate to each other. They relate to "is this branch likely to be taken or not".
Both can be likely.